// @squiddleton/fortnite-api/dist/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoints = void 0;
var Endpoints;
(function (Endpoints) {
    Endpoints["AES"] = "https://fortnite-api.com/v2/aes";
    Endpoints["Banners"] = "https://fortnite-api.com/v1/banners";
    Endpoints["BannerColors"] = "https://fortnite-api.com/v1/banners/colors";
    Endpoints["AllCosmetics"] = "https://fortnite-api.com/v2/cosmetics";
    Endpoints["NewCosmetics"] = "https://fortnite-api.com/v2/cosmetics/new";
    Endpoints["BRCosmetics"] = "https://fortnite-api.com/v2/cosmetics/br";
    Endpoints["Tracks"] = "https://fortnite-api.com/v2/cosmetics/tracks";
    Endpoints["Instruments"] = "https://fortnite-api.com/v2/cosmetics/instruments";
    Endpoints["Cars"] = "https://fortnite-api.com/v2/cosmetics/cars";
    Endpoints["LEGO"] = "https://fortnite-api.com/v2/cosmetics/lego";
    Endpoints["LEGOKits"] = "https://fortnite-api.com/v2/cosmetics/lego/kits";
    Endpoints["Beans"] = "https://fortnite-api.com/v2/cosmetics/beans";
    Endpoints["BRCosmeticById"] = "https://fortnite-api.com/v2/cosmetics/br/{cosmetic-id}";
    Endpoints["BRCosmeticSearch"] = "https://fortnite-api.com/v2/cosmetics/br/search";
    Endpoints["BRCosmeticsSearchAll"] = "https://fortnite-api.com/v2/cosmetics/br/search/all";
    Endpoints["BRCosmeticsSearchByIds"] = "https://fortnite-api.com/v2/cosmetics/br/search/ids";
    Endpoints["CreatorCode"] = "https://fortnite-api.com/v2/creatorcode";
    Endpoints["Map"] = "https://fortnite-api.com/v1/map";
    Endpoints["News"] = "https://fortnite-api.com/v2/news";
    Endpoints["BRNews"] = "https://fortnite-api.com/v2/news/br";
    Endpoints["STWNews"] = "https://fortnite-api.com/v2/news/stw";
    Endpoints["CreativeNews"] = "https://fortnite-api.com/v2/news/creative";
    Endpoints["Playlists"] = "https://fortnite-api.com/v1/playlists";
    Endpoints["PlaylistById"] = "https://fortnite-api.com/v1/playlists/{playlist-id}";
    Endpoints["Shop"] = "https://fortnite-api.com/v2/shop";
    Endpoints["BRStats"] = "https://fortnite-api.com/v2/stats/br/v2";
    Endpoints["BRStatsByAccountId"] = "https://fortnite-api.com/v2/stats/br/v2/{accountId}";
})(Endpoints = exports.Endpoints || (exports.Endpoints = {}));
